# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/canvas/OffscreenCanvas.cpp
# Commit: 904076df6327
# Full Hash: 904076df6327430699a83cfc75c2caca72dfa872
# Author: Ehsan Akhgari <ehsan@mozilla.com>
# Date: 2019-04-25 22:08:17
# Description:
#   Bug 1546390 - Enable determining whether the current document should respect resist fingerprinting mode on the main thread in OffscreenCanvas.toBlob(); r=baku
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D28593
# ==============================================================================

diff -r b9f3ca50c00e -r 904076df6327 dom/canvas/OffscreenCanvas.cpp
--- a/dom/canvas/OffscreenCanvas.cpp	Thu Apr 25 07:47:58 2019 +0000
+++ b/dom/canvas/OffscreenCanvas.cpp	Thu Apr 25 07:47:45 2019 +0000
@@ -251,13 +251,17 @@
 
   RefPtr<EncodeCompleteCallback> callback = new EncodeCallback(global, promise);
 
-  // TODO: Can we obtain the context and document here somehow
-  // so that we can decide when usePlaceholder should be true/false?
-  // See https://trac.torproject.org/18599
-  // For now, we always return a placeholder if fingerprinting resistance is on.
-  dom::WorkerPrivate* workerPrivate = dom::GetCurrentThreadWorkerPrivate();
-  bool usePlaceholder =
-      nsContentUtils::ShouldResistFingerprinting(workerPrivate->GetPrincipal());
+  bool usePlaceholder;
+  if (NS_IsMainThread()) {
+    nsCOMPtr<nsPIDOMWindowInner> window = do_QueryInterface(GetGlobalObject());
+    Document* doc = window->GetExtantDoc();
+    usePlaceholder =
+        doc ? nsContentUtils::ShouldResistFingerprinting(doc) : false;
+  } else {
+    dom::WorkerPrivate* workerPrivate = dom::GetCurrentThreadWorkerPrivate();
+    usePlaceholder = nsContentUtils::ShouldResistFingerprinting(
+        workerPrivate->GetPrincipal());
+  }
   CanvasRenderingContextHelper::ToBlob(aCx, global, callback, aType, aParams,
                                        usePlaceholder, aRv);
 